python/ray/includes/gcs_client.pxi

```diff
@@ -135,40 +135,149 @@ cdef class NewGcsClient:
+            CRayStatus status
+        with nogil:
+            status = self.inner.get().InternalKV().Exists(
+                ns, key, timeout_ms, opt_exists.value())
+        return raise_or_return(convert_optional_bool(status, opt_exists))
+
+    #############################################################
+    # Internal KV async methods
+    #############################################################
+
+    def async_internal_kv_get(
+        self, c_string key, namespace=None, timeout=None
+    ) -> Future[Optional[bytes]]:
+        cdef:
+            c_string ns = namespace or b""
+            int64_t timeout_ms = round(1000 * timeout) if timeout else -1
+            fut = incremented_fut()
+            void* fut_ptr = <void*>fut
         with nogil:
             check_status_timeout_as_rpc_error(
-                self.inner.get().InternalKV().Exists(ns, key, timeout_ms, exists)
-            )
-        return exists
+                self.inner.get().InternalKV().AsyncInternalKVGet(
+                    ns, key, timeout_ms,
+                    OptionalItemPyCallback[c_string](
+                        convert_optional_str_none_for_not_found,
+                        assign_and_decrement_fut,
+                        fut_ptr)))
+        return asyncio.wrap_future(fut)
+
+    def async_internal_kv_multi_get(
+        self, keys: List[bytes], namespace=None, timeout=None
+    ) -> Future[Dict[bytes, bytes]]:
+        cdef:
+            c_string ns = namespace or b""
+            int64_t timeout_ms = round(1000 * timeout) if timeout else -1
+            c_vector[c_string] c_keys = [key for key in keys]
+            fut = incremented_fut()
+            void* fut_ptr = <void*>fut
+        with nogil:
+            check_status_timeout_as_rpc_error(
+                self.inner.get().InternalKV().AsyncInternalKVMultiGet(
+                    ns, c_keys, timeout_ms,
+                    OptionalItemPyCallback[unordered_map[c_string, c_string]](
+                        convert_optional_multi_get,
+                        assign_and_decrement_fut,
+                        fut_ptr)))
+        return asyncio.wrap_future(fut)
+
+    def async_internal_kv_put(
+        self, c_string key, c_string value, c_bool overwrite=False, namespace=None,
+        timeout=None
+    ) -> Future[int]:
+        # TODO(ryw): the sync `internal_kv_put` returns bool while this async version
+        # returns int. We should make them consistent.
+        cdef:
+            c_string ns = namespace or b""
+            int64_t timeout_ms = round(1000 * timeout) if timeout else -1
+            fut = incremented_fut()
+            void* fut_ptr = <void*>fut
+        with nogil:
+            check_status_timeout_as_rpc_error(
+                self.inner.get().InternalKV().AsyncInternalKVPut(
+                    ns, key, value, overwrite, timeout_ms,
+                    OptionalItemPyCallback[int](
+                        convert_optional_int,
```

Comment:
In theory, this is never optional right?

Reply:
If status is not OK, the optional[int] would be nullopt. Otherwise, right it should never be optional.

Reply:
added assert